### PR TITLE
Implement library

### PR DESCRIPTION
**Pull request by aksullivan86**
Implements the library. Note that there should be test coverage for the acceptance criteria given. I also have the tests configured in github actions, so you can see the result of their runs on every commit.

### CODE REVIEW

File: jail/cgroup.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+					Rate:  j.CgroupLimits.DeviceRate,
+				},
+			},
+		},
+	}
+
+	return resources, nil
+}
+
+// Helper function for a consistent cgroup name from a jobId.
```
Comment (tigrato) [inline jail/cgroup.go:63]: use golang default format for comments:

// {functionName} ...

File: jail/executor.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/docker/docker/pkg/reexec"
+)
+
+// Start the command specified in Jail on the host system. Before execution, a cgroup
+// is created. Limits that you specify in CgroupLimits are then applied to the cgroup.
+// Finally before the process is actually started, it is added to the cgroup and then
+// ran. Callers are responsible for calling Wait() on the returned *exec.Cmd to know
+// when the process has completed. The paramater writer is where stdout and stderr
+// output of the process will ship.
+func (j *Jail) Start(writer io.Writer) (*exec.Cmd, error) {
```
Comment (tigrato) [inline jail/executor.go:19]: instead of receiving a `writer` can we receive a `exec.Cmd` ?
I think it would be much easier to understand what is happening and the responsibility of the `exec.Cmd` is clear.

What I would do:

- receive a cmd we want to run
- re-setup stdin  (if provided you can merge multi readers with `io.MultiReader` and replace the stdin)
- Start
- Send payload
- close the initial stdin reader and let the MultiReader handle the concatenation.

File: jail/reexec.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	}
+
+	myPid := os.Getpid()
+	if err := control.AddProc(uint64(myPid)); err != nil {
+		fmt.Fprintf(os.Stderr, "failed adding PID %d to cgroup: %v\n", myPid, err)
+		os.Exit(1)
+	}
+
+	//fmt.Printf("reexec'd with params: %s\npid: %d\n", params, myPid)
+	cmd := exec.Command(params.Cmd, params.CmdArgs...)
```
Comment (tigrato) [inline jail/reexec.go:39]: Is it possible to avoid exec twice and create two different PIDs?

If the reexec PID dies, the children's PID will be reparented to the primary process that does not have limits.

File: job/stream.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+func (jls *JobLogStream) Read(p []byte) (int, error) {
+	for {
+		// wait for start
+		jls.ConsumerMutex.RLock()
+		if jls.started {
+			jls.ConsumerMutex.RUnlock()
+			break
+		}
+		jls.ConsumerMutex.RUnlock()
+		time.Sleep(200 * time.Millisecond)
```
Comment (tigrato) [inline job/stream.go:48]: is it possible to avoid sleeps?

File: job/stream.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+		ConsumerMutex: &sync.RWMutex{},
+		Buffer:        bytes.Buffer{},
+		Condition:     sync.NewCond(&sync.Mutex{}),
+		mutex:         &sync.Mutex{},
+	}
+	stream.Condition.L.Lock()
+
+	go func() {
+		for {
+			if job.Completed() {
```
Comment (tigrato) [inline job/stream.go:23]: Why do you need this wake-up?

What if the writer function wakes the `sync.Cond` on writes and when the stream completes?

File: job/stream.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+		jls.mutex.Unlock()
+		jls.ConsumerMutex.Unlock()
+		jls.Condition.Broadcast() // wake up
+	}()
+
+	return jls.Buffer.Write(p)
+}
+
+// True when no additional data will ever come in (aka when the Job is complete)
+func (jls *JobLogStream) NoNewDataEver() bool {
```
Comment (tigrato) [inline job/stream.go:87]: the function was more clear if the name was:  `JobCompleted`

File: job/stream.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+			// wait for more input
+			return n, nil
+		}
+	}
+
+	return n, err
+}
+
+// Writes data to the Buffer for readers. Notifies any readers of the write.
+func (jls *JobLogStream) Write(p []byte) (int, error) {
```
Comment (tigrato) [inline job/stream.go:73]: I would avoid making this function available outside of the lib. 
Use an interface to at least hide this function since it's not supposed to be called by anyone else except this lib

File: job/job.go [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+	if err := j.updateStateWithLock(JobStateRunning); err != nil {
+		return err
+	}
+
+	j.mutex.Lock()
+	var err error
+	if util.IsRunningWithinCi() {
+		// skip cgroup/jail task(s)
+		j.cmd = exec.CommandContext(j.ctx, j.CmdStr, j.CmdStrArgs...)
+		j.cmd.Stdout = io.MultiWriter(j)
```
Comment (nklaassen) [inline job/job.go:170]: `MultiWriter` is not necessary here

File: worker/worker.go [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+	}
+
+	return nil
+}
+
+// Reads from your JobLogStreamer, doing with the data as you specify in your consumer.
+// Must have previously called JobLogsCapture().
+func (w *Worker) JobLogs(
+	jls *job.JobLogStream,
+	consumer func(data []byte) error,
```
Comment (nklaassen) [inline worker/worker.go:104]: Where do you get the JobLogStream to pass to JobLogsCapture so you can then call JobLogs? This API is quite confusing to use and all the data is internally copied many times. Could you simplify? Small interfaces like `io.Write(Closer)` and `io.Read(Closer)` can be very simple and useful for tasks like this

File: jail/reexec.go [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+// This should only be ran by reexec. As this is a forked process, we can safely add this process
+// to the cgroup since it has its own process ID. Remember, this is called as a main function, so
+// to signify error, you can simply write to stdout/stderr, and exit non zero.
+func jobCmdRunnerMain() {
+	// Prevent other go routines from executing in this thread. Never release so when this function
+	// completes, so does the thread.
+	runtime.LockOSThread()
+
+	var params *JailParams
+	if err := json.NewDecoder(os.Stdin).Decode(&params); err != nil {
```
[the remark this answers is not in the document]
Comment (aksullivan86) [reply jail/reexec.go:20]: ah yes, that was me being hasty I think

File: jail/executor.go [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+	j.control = control
+
+	recmd := reexec.Command("job-cmd-runner")
+
+	w, err := recmd.StdinPipe()
+	if err != nil {
+		return nil, fmt.Errorf("reexec.StdinPipe() failed: %v", err)
+	}
+
+	recmd.Stdout = io.MultiWriter(writer)
```
[the remark this answers is not in the document]
Comment (aksullivan86) [reply jail/executor.go]: 642eb296e989ad4dee7ae91e216122158c627dfc

File: job/job.go (hunk 8 above)
Comment (aksullivan86) [reply job/job.go]: 642eb296e989ad4dee7ae91e216122158c627dfc

File: job/stream.go [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+		if jls.started {
+			jls.ConsumerMutex.RUnlock()
+			break
+		}
+		jls.ConsumerMutex.RUnlock()
+		time.Sleep(200 * time.Millisecond)
+	}
+
+	if !jls.NoNewDataEver() {
+		jls.Condition.Wait() // wait for writer
```
[the remark this answers is not in the document]
Comment (aksullivan86) [reply job/stream.go:52]: https://pkg.go.dev/sync#Cond.Wait
```
Each Cond has an associated Locker L (often a *Mutex or *RWMutex), which must be held when changing the condition and when calling the Wait method. 
```
This does make it sound like that, though I am surprised it would be required. However if I add a `jls.Condition.L.Lock()` above that wait, when log streams happen things just lock forever. This is something I would have spent more time on for a production quality release testing the streaming and all edge cases.

File: job/stream.go (hunk 6 above)
Comment (aksullivan86) [reply job/stream.go:87]: the idea is that the job only completed alone isn't enough. You can still get logs for a completed job. Thats why i didnt use that name.

File: worker/worker.go (hunk 9 above)
Comment (aksullivan86) [reply worker/worker.go:104]: As it stands that would be in the responsibility of the server, for example [here](https://github.com/aksullivan86/job-worker/pull/3/files#diff-53cd25fc908d5e63cc8675afbd4ff0910572a985a7cfec27f4460eb674f3b98eR237).

I will take a pass at perhaps simplifying the log streaming api time permitting. I'm aware that was the weakest point, and something I have the least experience in. Though it does actually work from my tests. I totally agree with you guys about the suggestions you are making for improvements on the streaming though. Im honestly not sure I will have time to implement the redesign however by deadline :(

File: job/stream.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (nklaassen) [reply job/stream.go:52]: It probably deadlocks because you call `stream.Condition.L.Lock()` in `NewJobLogStream`. `sync.Cond` is more complicated than it seems and almost everyone who tries to use it in this interview gets it wrong. The question to think about is how can you guarantee that you haven't missed a broadcast right before you called `Wait()`? But even better than using `sync.Cond` correctly would be to avoid using it at all in my opinion

File: job/job.go [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+
+// Like a normal Write(), but the output is also written to Job.output. Job.output is used for
+// log retrieval.
+func (j *Job) Write(p []byte) (int, error) {
+	j.mutex.Lock()
+	defer j.mutex.Unlock()
+
+	j.output = append(j.output, p...)
+
+	return j.writer.Write(p)
```
[the remark this answers is not in the document]
Comment (aksullivan86) [reply job/job.go]: Good catch! That was from a previous implementation I was trying and failed to remove later :( Corrected c80369a56846a3ed66a894b937476e4ab2719da1

File: jail/cgroup.go (hunk 1 above)
Comment (aksullivan86) [reply jail/cgroup.go]: 65ca28edf0e8a6de384be19cf93098ffa360800f

File: jail/init.go [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+)
+
+func init() {
+	reexec.Register("job-cmd-runner", jobCmdRunnerMain)
+	if reexec.Init() {
+		os.Exit(0)
+	}
+}
+
+func NewJail(params *JailParams, cgroupLimits *CgroupLimits) *Jail {
```
[the remark this answers is not in the document]
Comment (aksullivan86) [reply jail/init.go:18]: 65ca28edf0e8a6de384be19cf93098ffa360800f
